server/src/main/java/org/elasticsearch/index/mapper/ConstantFieldType.java

```diff
@@ -79,6 +79,10 @@ private static String valueToString(Object value) {
     public final Query termQuery(Object value, QueryShardContext context) {
         String pattern = valueToString(value);
         if (matches(pattern, context)) {
+            if (context != null && context.getMapperService().hasNested()) {
+                // type filters are expected not to match nested docs
+                return Queries.newNonNestedFilter(context.indexVersionCreated());
```

Comment:
See comment in tests below, I wasn't sure if on 7.x we still need this so left it here for discussion.